packages/yarnpkg-core/sources/ConfigurableReport.ts

```diff
@@ -0,0 +1,80 @@
+import {Configuration}                     from './Configuration';
+import {MessageName, stringifyMessageName} from './MessageName';
+import {Report}                            from './Report';
+
+export enum LogLevel {
+  ERROR = `error`,
+  WARNING = `warning`,
+  INFO = `info`,
+  DISCARD = `discard`,
+}
+
+type FindLogLevel = (name: MessageName | null, text: string, defaultLevel: LogLevel) => LogLevel;
+
+const useDefaultLogLevel: FindLogLevel = (name, text, level) => level;
+
+export abstract class ConfigurableReport extends Report {
+  private findLogLevel: FindLogLevel;
+
+  constructor(configuration: Configuration) {
+    super();
+
+    const logFilter = configuration.get(`logFilter`);
+
+    if (logFilter.size === 0) {
+      this.findLogLevel = useDefaultLogLevel;
+    } else {
+      const logFilterByRegExp = Array.from(logFilter)
+        .filter(([key]) => !/^YN\d{4}$/.test(key))
+        .map(([key, level]) => [new RegExp(key), level] as const);
+
+      this.findLogLevel = (name, text, defaultLevel) => {
+        if (name != null) {
+          const levelByName = logFilter.get(stringifyMessageName(name));
+
+          if (typeof levelByName !== `undefined`) {
+            return levelByName ?? defaultLevel;
+          }
+        }
+
+        for (const [regExp, logLevel] of logFilterByRegExp) {
+          if (regExp.test(text)) {
+            return logLevel ?? defaultLevel;
+          }
+        }
+
+        return defaultLevel;
+      };
+    }
+  }
+
+  abstract writeReportInfo(name: MessageName | null, text: string): void;
+  abstract writeReportWarning(name: MessageName, text: string): void;
+  abstract writeReportError(name: MessageName, text: string): void;
+
+  writeReportLevel(level: LogLevel, name: MessageName | null, text: string): void {
+    switch (this.findLogLevel(name, text, level)) {
```

Comment:
What do you think about preventing to hide when `name` is either `UNNAMED` or `null`? On our side it'll give us some incentive to provide accurate message names, and on the user's side it means they'll have less risks to accidentally hide important messages (either because of the layout, or just because it'd be part of the expected command output).

Reply:
Hmm, yeah that's a real danger when filtering on message content 🤔 
On the other hand, what you propose would make it impossible to hide messages from 3rd party plugins, as those don't have names.

Maybe we need to provide a way for plugins to register names?
Those wouldn't be `YNxxxx` but e.g. `Exxyyy` where `xx` are based on the plugin and `yyy` are a number passed into the logger by the plugin.
The plugin could map message ids onto more information or even a URL (e.g. to a website, a markdown document on github, a title in the README, etc) via a hook. The yarn website would need to understand these 3rd party messages and tell the user to run a command to get more info wrt the message.

Reply:
> Maybe we need to provide a way for plugins to register names?

Interestingly the documentation references this ability [here](https://yarnpkg.com/advanced/error-codes):

> Are you a plugin author and want to declare your own error codes that don't match the semantic of the ones provided here? Please relinquish one character and use the YNX prefix (ex YNX001) instead of YN0!

But this suggestion doesn't work at the moment since the functions take an integer, not a string. Perhaps we could let the report functions accept a `[pluginId, messageId]` tuple as valid message name, and print it with the format you suggest?

Reply:
> Perhaps we could let the report functions accept a `[pluginId, messageId]` tuple as valid message name, and print it with the format you suggest?

I had a couple of options in mind, ranging from something like a `[pluginId, messageId]` tuple as `name` parameter for the logging functions to giving plugins specific report instances for that plugin which would automatically map message ids onto codes for that plugin. The latter could prevent plugins from (ab)using the `YNxxxx` message names, but it would lead to a more complicated implementation (what if such a plugin reporter gets passed into a core function?)

I'm currently leaning towards providing a way for plugins to define their own message "enum", e.g.

```ts
import {createPluginMessageNames} from '@yarnpkg/core';

// typeof AngularPluginMessageName = Readonly<{
//   ANGULAR_CLI_NOT_FOUND: PluginMessageName;
// }>
export const AngularPluginMessageName = createPluginMessageNameEnum('@yarnpkg/plugin-angular', {
  ANGULAR_CLI_NOT_FOUND: 1,
});

// using it:

// Report functions accept @yarnpkg/core's MessageName, null,
// or PluginMessageName as message name
report.reportError(AngularPluginMessageName.ANGULAR_CLI_NOT_FOUND, `${
  structUtils.prettyIdent(configuration, structUtils.makeIdent('angular', 'cli'))
} is not installed in your project`);

// this API doesn't depend on what PluginMessageName actually is,
// it could e.g. be a [pluginId, messageId] tuple
```

Reply:
If you need someone to test this out on the plugin side I would be willing to help out on that.